README
NAME
    Catalyst::ControllerRole::Resources::Verbs - Additional verbs for
    Catalyst::Controller::Resources

DESCRIPTION
    Catalyst::Controller::Resources makes it easy to create a
    resource-oriented controller, but the actual mechanics of extending it
    with additional "verbs" is a little opaque. This package provides a
    parameterized role designed to make adding additional verbs easy, as
    well as a number of concrete roles implementing additional common
    actions.

PARAMETERIZED ROLE
    See Catalyst::ControllerRole::Resources::Verb for details.

CONCRETE ROLES
    These implementations live under the
    Catalyst::ControllerRole::Resources::Verb::* namespace.

  Count
    Catalyst::ControllerRole::Resources::Verb::Count implements a COLLECTION
    action that is expected to return the collection count (whatever that
    means in the context you're using it).

        path: count
        args: none
        method: GET

SEE ALSO
    Catalyst::Controller::Resources, Catalyst::Controller::REST.

BUGS AND LIMITATIONS
    There are no known bugs in this module.

    Please report problems to Chris Weyl <cweyl@alumni.drew.edu>, or
    (preferred) to this package's RT tracker at
    <bug-Catalyst-ControllerRole-Resources-Verbs@rt.cpan.org>.

    Patches are welcome.

AUTHOR
    Chris Weyl <cweyl@alumni.drew.edu>

LICENSE AND COPYRIGHT
    Copyright (c) 2010 Chris Weyl <cweyl@alumni.drew.edu>

    This library is free software; you can redistribute it and/or modify it
    under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2.1 of the License, or (at
    your option) any later version.

    This library is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
    General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this library; if not, write to the

        Free Software Foundation, Inc.
        59 Temple Place, Suite 330
        Boston, MA  02111-1307  USA



ChangeLog
Changes for Catalyst::ControllerRole::Resources::Verbs

0.001_01 - Sun Jun 13 2010
	* Initial devel release; working but no real test suite yet



lib/Catalyst/ControllerRole/Resources/Verbs.pm
#############################################################################
#
# Author:  Chris Weyl (cpan:RSRCHBOY), <cweyl@alumni.drew.edu>
# Company: No company, personal work
#
# Copyright (c) 2010  <cweyl@alumni.drew.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
#############################################################################

package Catalyst::ControllerRole::Resources::Verbs;

our $VERSION = '0.001_01';

1;

__END__

=head1 NAME

Catalyst::ControllerRole::Resources::Verbs - Additional verbs for Catalyst::Controller::Resources

=head1 DESCRIPTION

L<Catalyst::Controller::Resources> makes it easy to create a resource-oriented
controller, but the actual mechanics of extending it with additional "verbs"
is a little opaque.  This package provides a parameterized role designed to
make adding additional verbs easy, as well as a number of concrete roles
implementing additional common actions.

=head1 PARAMETERIZED ROLE

See L<Catalyst::ControllerRole::Resources::Verb> for details.

=head1 CONCRETE ROLES

These implementations live under the
Catalyst::ControllerRole::Resources::Verb::* namespace.

=head2 Count

L<Catalyst::ControllerRole::Resources::Verb::Count> implements a COLLECTION
action that is expected to return the collection count (whatever that means in
the context you're using it).

    path: count
    args: none
    method: GET

=head1 SEE ALSO

L<Catalyst::Controller::Resources>, L<Catalyst::Controller::REST>.

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems to Chris Weyl <cweyl@alumni.drew.edu>, or (preferred)
to this package's RT tracker at
<bug-Catalyst-ControllerRole-Resources-Verbs@rt.cpan.org>.

Patches are welcome.

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>


=head1 LICENSE AND COPYRIGHT

Copyright (c) 2010 Chris Weyl <cweyl@alumni.drew.edu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the

    Free Software Foundation, Inc.
    59 Temple Place, Suite 330
    Boston, MA  02111-1307  USA

=cut




lib/Catalyst/ControllerRole/Resources/Verb/Count.pm
package Catalyst::ControllerRole::Resources::Verb::Count;

use Moose::Role;
use namespace::autoclean;

our $VERSION = '0.001_01';

with 'Catalyst::ControllerRole::Resources::Verb' => {
    target  => 'collection',
    verb    => 'count',
    method  => 'GET',
    path    => 'count',
};

1;

__END__

=head1 NAME

Catalyst::ControllerRole::Resources::Verb::Count

=head1 SYNOPSIS

    # in your controller
    use Moose;
    use namespace::autoclean;

    BEGIN { extends 'Catalyst::Controller::Resources' }
    with 'Catalyst::ControllerRole::Resources::Verb::Count';

    sub count { ... }

=head1 DESCRIPTION

L<Catalyst::ControllerRole::Resources::Verb::Count> implements a COLLECTION
action that is expected to return the collection count (whatever that means in
the context you're using it).

    path: count
    args: none
    method: GET

=head1 SEE ALSO

L<Catalyst::ControllerRole::Resources::Verb>,
L<Catalyst::ControllerRole::Resources::Verbs>

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2010 Chris Weyl <cweyl@alumni.drew.edu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the

    Free Software Foundation, Inc.
    59 Temple Place, Suite 330
    Boston, MA  02111-1307  USA

=cut



lib/Catalyst/ControllerRole/Resources/Verb.pm
#############################################################################
#
# Author:  Chris Weyl (cpan:RSRCHBOY), <cweyl@alumni.drew.edu>
# Company: No company, personal work
#
# Copyright (c) 2010  <cweyl@alumni.drew.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
#############################################################################

package Catalyst::ControllerRole::Resources::Verb;

use MooseX::Role::Parameterized;
use namespace::autoclean;

our $VERSION = '0.001_01';

# FIXME should type-validate target, verb (at least)
parameter target => (isa => 'Str', required => 1);
parameter verb   => (isa => 'Str', required => 1);
parameter method => (isa => 'Str', required => 1);
parameter path   => (isa => 'Str', required => 1);

my %attribute = (
    member     => '_additional_member_actions',
    collection => '_additional_collection_actions',
);

role {
    my $p = shift @_;

    die $p->target . ' is not a valid target!'
        unless exists $attribute{$p->target};

    requires $p->verb;

    around $attribute{$p->target} => sub {
        my ($orig, $self) = (shift, shift);
        my $actions = $self->$orig(@_);

        $actions->{$p->verb} = { method => $p->method, path => $p->path };
        return $actions;
    };
};

1;

__END__

=head1 NAME

Catalyst::ControllerRole::Resources::Verb

=head1 SYNOPSIS

    use Moose;
    use namespace::autoclean;

    BEGIN { extends 'Catalyst::Controller::Resources' }

    with 'Catalyst::ControllerRole::Resources::Verb' => {
        target  => 'collection',
        verb    => 'count',
        method  => 'GET',
        path    => 'count',
    };

    # make sure to implement the verb...
    sub count { ... }

    # profit!

=head1 DESCRIPTION

This is a parameterized Moose role, that handles the task of extending
L<Catalyst::Controller::Resources> with additional "verbs".

It takes 4 mandatory arguments:

=over 4

=item B<target> - 'member' or 'collection'

=item B<verb> - implementing controller method name

=item B<method> - HTTP method; e.g. GET, POST, etc

=item B<path> - the pathpart of the action

=back

=head1 SEE ALSO

L<Catalyst::ControllerRole::Resources::Verbs>

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2010 Chris Weyl <cweyl@alumni.drew.edu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the

    Free Software Foundation, Inc.
    59 Temple Place, Suite 330
    Boston, MA  02111-1307  USA

=cut


